draft-ietf-ppm-dap.md

```diff
@@ -1554,14 +1580,29 @@ struct {
 where `prepare_resps` are the outbound prep steps computed in the previous step.
 The order MUST match `AggregationJobInitReq.prepare_inits`.
 
-The Helper responds to the Leader with HTTP status code 201 Created and a body
+If the Helper is asynchronously processing the aggregation job, the Leader makes
+HTTP GET requests to
+`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step=0` to check
+on the status of the aggregation job. If the aggregation job is not finished yet,
+the Helper MAY wait to respond until the job is complete. Once the aggregation
+job is ready, or if it is already ready, the Helper responds to the GET request
+with HTTP status 200 OK and a body consisting of the `AggregationJobResp`, with
+media type "application/dap-aggregation-job-resp". If the aggregation job is not
+finished yet, and the Helper chooses not to wait for the job to finish, the
+Helper responds with HTTP status 202 Accepted. The Helper's response SHOULD
+include a Retry-After header field to suggest a polling interval to the Leader.
+
+Otherwise, if the Helper is synchronously processing the aggregation job, the
+Helper responds to the PUT request with HTTP status 201 Created and a body
 consisting of the `AggregationJobResp`, with media type
 "application/dap-aggregation-job-resp".
 
-Changing an aggregation job's parameters is illegal, so further requests to
-`PUT /tasks/{tasks}/aggregation_jobs/{aggregation-job-id}` for the same
-`aggregation-job-id` but with a different `AggregationJobInitReq` in the body
-MUST fail with an HTTP client error status code.
+Changing an aggregation job's parameters is illegal, so further HTTP PUT
+requests to `/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for the
+same `aggregation-job-id` but with a different `AggregationJobInitReq` in the
+body MUST fail with an HTTP client error status code. For further requests with
+the same `AggregationJobInitReq` in the body, the Helper MUST respond with HTTP
+status 201 Created.
```

Comment:
I think idempotency is a required property of a [PUT request](https://www.rfc-editor.org/rfc/rfc9110#section-9.3.4-2), rather than related to skew recovery.

We could make it a SHOULD. The helper's behavior then should respond with an error code though, I don't think it's valid for the helper to accept more than one valid `AggregationJobInitReq`.

Reply:
I've turned it into a SHOULD. I suppose for many-step VDAFs, the helper could want to discard its response to `AggregationJobInitReq` as later steps take place, i.e. `n-2, n-1, n`, `n-2` may be zero and can be discarded/overwritten while still supporting skew recovery. In which case the helper would respond with something like 410 Gone.

Reply:
> Not for this PR, but perhaps that section should be generalized, since the initialization request is effectively "step 0".

I'm hoping in the next PR that section can be stricken entirely, now that we have step encoded in the URI. Although I haven't thought about it in the face of supporting both sync/async.